### ISSUE TITLE
Noexcept `throw()` specifier removed in C++20

### ISSUE THREAD

**Issue by dmitrykobets-msft** [llvm/llvm-project#58946]
https://en.cppreference.com/w/cpp/language/noexcept_spec mentions that the `throw()` exception specifier was deprecated in C++17 and removed in C++20. This doesn't seem to be the case when compiling with clang:
```cpp
void foo() throw() {} // no diagnostic issued
```

**Comment by shafik** [comment]
So you can obtain a warning w/ `-Wdeprecated-dynamic-exception-spec` but it is not on by default which seems like a bug.

Note, it looks like neither gcc nor MSVC warn either: https://godbolt.org/z/GjqoY7xTW

**Comment by joeywatts** [comment]
I'd be willing to take this one! Should it be a new error diagnostic like "empty dynamic exception specifications were removed in C++20"?

**Comment by shafik** [comment]
I don't believe we need a new diagnostic, we just need `Wdeprecated-dynamic-exception-spec` to be enabled by default. I think we just need to remove the `DefaultIgnore` on the diagnostic `warn_exception_spec_deprecated`

**Comment by joeywatts** [comment]
Gotcha, makes sense. Is it okay that it's still a warning and not an error when targeting C++20?

**Comment by zygoloid** [comment]
The minimum requirement is that it's a warning under `-pedantic` and an error under `-pedantic-errors`. Clang's strictly conforming mode is enabled by `-pedantic`, and conformance to the C++ standard requires that we produce "at least one diagnostic" for invalid input -- it's enough for us to warn and accept anyway. `-pedantic-errors` is supposed to reject all non-standard code with an error.

The default mode (without `-pedantic`) can have enabled-by-default extensions, and I don't think we have any firm rules as to whether extensions like this one (accepting valid C++17 code that is meaningful but not valid in C++20) should be warnings by default or not. Warning by default is probably the right thing to do here, unless we think that's going to be bad for our users and they'll routinely just disable this warning and leave it disabled. (Given that there's a backwards-compatible source change that resolves the warning, I'd hope that they won't!)

**Comment by AaronBallman** [comment]
I think our policy should be to always diagnose deprecated functionality by default unless there is a strong rationale for silencing it (stronger than "it might be disruptive" -- it needs to be a concrete problem rather than a hypothetical). The standards committees generally expect that deprecated functionality is something that is being actively diagnosed so that the design space can be eventually retaken. Failing to diagnose by default means users have to remember to opt-in and that makes removing the deprecated functionality significantly harder.

In terms of this particular extension, I don't have a problem with continuing to support `throw()` in C++20 and later (with diagnostic), but I think I'd want some mention in the release notes/documentation about our future plans for it. Unless there are system headers using this construct in C++20 and later, I think we don't want to commit to this extension forever, so I think it's reasonable to add information about the potential for it to be upgraded to an error in the future (we might want to go with warning-defaults-to-error as another step before we outright make it an error).
